akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -699,18 +790,63 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
-  final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
+  final protected def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
     new AsyncCallback[T] {
       override def invoke(event: T): Unit =
         interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
     }
   }
 
+  // TODO rethink if it's worth it; move the Updater to an object; etc
+  // via atomics since users may do something weird - try to Future { init } or something...
+  private final val _setStageActorRef = AtomicReferenceFieldUpdater.newUpdater(classOf[GraphStageLogic], classOf[StageActorRef], "_stageActorRefDoNotCallMeDirectly")
+  @volatile private var _stageActorRefDoNotCallMeDirectly: StageActorRef = _ // DO NOT TOUCH!
+  final implicit def stageActorRef: ActorRef =
+    if (_stageActorRefDoNotCallMeDirectly ne null) _stageActorRefDoNotCallMeDirectly // TODO rethink this
+    else throw StageActorRefNotInitializedException()
+
+  case class StageActorRefNotInitializedException() extends RuntimeException("You must first call getStageActorRef, to initialize the Actors behaviour")
```

Comment:
I’d vote for removing all this, getting rid of the additional field and letting users store the ref they get back within their logic—that also places the cost where the demand is.